core/primitives/src/transaction.rs

```diff
@@ -220,6 +226,40 @@ impl From<DeleteAccountAction> for Action {
     }
 }
 
+#[cfg_attr(feature = "deepsize_feature", derive(deepsize::DeepSizeOf))]
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
+pub struct DelegateAction {
+    pub reciever_id: AccountId,
+    pub deposit: Balance,
+    pub nonce: u64,
+    pub actions: Vec<Action>,
+}
+#[cfg_attr(feature = "deepsize_feature", derive(deepsize::DeepSizeOf))]
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
+pub struct SignedDelegateAction {
+    // Borsh doesn't support recursive types. Therefore this field
+    // is deserialized to DelegateAction in runtime
+    pub delegate_action_serde: Vec<u8>,
```

Comment:
I don't fully understand why not just use delegate_action: DelegateAction here?
See SignedTransaction for example - https://github.com/near/nearcore/blob/62f66348ae7871e2c5879a60395f4003a9e9bbf1/core/primitives/src/transaction.rs#L227

You can add borsh_init to initialize whatever data is needed.

Reply:
>I don't fully understand why not just use delegate_action: DelegateAction here?
The comment is wrong. The issue is that there is a type recursion:
 ```
pub struct Transaction {
    ....
    pub actions: Vec<Action>, // <--- Recursion
}

pub enum Action {
    ...
    Delegate(SignedDelegateAction),
}

pub struct SignedDelegateAction {
    pub delegate_action_serde: Vec<u8>,
    ....
}

pub struct DelegateAction {
    ...
    pub actions: Vec<Action>, /// <--- Recursion
}
```